File: crates/core/src/schemes/eip191.rs
//! secp256k1 + EIP-191 program: identity is the 20-byte Ethereum address;
//! the digest is wrapped in the EIP-191 personal-sign envelope before
//! signing/recovery.

use k256::ecdsa::{signature::hazmat::PrehashSigner, SigningKey as Secp256k1SigningKey};
use sha3::{Digest as Sha3Digest, Keccak256};
use solana_address::{address, Address};
use solana_instruction::Instruction;

use crate::instructions::{create_advance_instruction, create_initialize_instruction};
use crate::protocol::advance_vector_digest;
use crate::scheme::Scheme;

pub const EIP191_ETH_ADDRESS_LEN: usize = 20;

/// secp256k1 ECDSA + EIP-191 envelope — identity is the 20-byte ETH address.
pub const EIP191: Scheme = Scheme {
    program_id: address!("G6okL1MvXx7k5eytY7wRXNupXyYG1QVZW37ygAjMiTTu"),
    signature_len: 65,
    identity_len: EIP191_ETH_ADDRESS_LEN,
    stored_identity_len: EIP191_ETH_ADDRESS_LEN,
};

/// Derive the 20-byte Ethereum address from an uncompressed secp256k1 public
/// key. Accepts the 65-byte `0x04 || x || y` form or the raw 64-byte point.
pub fn eth_address_from_pubkey(uncompressed: &[u8]) -> [u8; EIP191_ETH_ADDRESS_LEN] {
    let point = match uncompressed.len() {
        65 => &uncompressed[1..],
        64 => uncompressed,
        _ => panic!("invalid uncompressed public key length"),
    };
    let hash: [u8; 32] = Keccak256::digest(point).into();
    let mut addr = [0u8; EIP191_ETH_ADDRESS_LEN];
    addr.copy_from_slice(&hash[12..32]);
    addr
}

/// 20-byte Ethereum address (the identity) for an EIP-191 secp256k1 key.
pub fn secp256k1_eip191_eth_address(
    signing_key: &Secp256k1SigningKey,
) -> [u8; EIP191_ETH_ADDRESS_LEN] {
    let verifying_key = signing_key.verifying_key();
    let uncompressed = verifying_key.to_encoded_point(false);
    eth_address_from_pubkey(uncompressed.as_bytes())
}

/// Initialize an EIP-191 vector account. `eth_address` is the 20-byte ETH
/// address (no padding).
pub fn create_initialize_secp256k1_eip191(
    payer: &Address,
    eth_address: &[u8; EIP191_ETH_ADDRESS_LEN],
) -> Instruction {
    create_initialize_instruction(payer, &EIP191, eth_address, eth_address)
}

/// `keccak256("\x19Ethereum Signed Message:\n32" || digest)` — the EIP-191
/// personal-sign envelope the on-chain program reproduces before
/// `secp256k1_recover`.
fn eip191_envelope_hash(digest: &[u8; 32]) -> [u8; 32] {
    let mut hasher = Keccak256::new();
    hasher.update(b"\x19Ethereum Signed Message:\n32");
    hasher.update(digest);
    hasher.finalize().into()
}

/// Sign the advance digest with an EIP-191 secp256k1 key, returning the
/// advance ix alone. Any CPI passthrough must be built separately via
/// [`crate::instructions::create_passthrough_instruction`] and included
/// among `pre_instructions` or `post_instructions` so the digest commits
/// to its bytes.
pub fn sign_advance_instruction_secp256k1_eip191(
    signing_key: &Secp256k1SigningKey,
    nonce: &[u8; 32],
    pre_instructions: &[Instruction],
    post_instructions: &[Instruction],
) -> Instruction {
    let identity = secp256k1_eip191_eth_address(signing_key);
    let digest = advance_vector_digest(
        &EIP191,
        nonce,
        &identity,
        pre_instructions,
        post_instructions,
    );
    let eth_digest = eip191_envelope_hash(&digest);
    let (sig, recid) = signing_key
        .sign_prehash(&eth_digest)
        .expect("secp256k1 signing failed");
    let mut sig_bytes = [0u8; 65];
    sig_bytes[..64].copy_from_slice(&sig.to_bytes());
    sig_bytes[64] = recid.to_byte();
    create_advance_instruction(&EIP191, &identity, &sig_bytes)
}

// ---------------------------------------------------------------------------
// Eip191 struct — implements SchemeMeta + Signer + Verifier
// ---------------------------------------------------------------------------

use k256::ecdsa::{RecoveryId, Signature, SigningKey, VerifyingKey};

use crate::scheme::{SchemeMeta, Signer, Verifier};

/// EIP-191 secp256k1 signer/verifier. Identity is the 20-byte Ethereum address.
#[derive(Clone)]
pub struct Eip191 {
    key: SigningKey,
}

impl Eip191 {
    pub fn from_seed(seed: &[u8; 32]) -> Self {
        Self {
            key: SigningKey::from_slice(seed).expect("valid secp256k1 scalar"),
        }
    }
}

impl SchemeMeta for Eip191 {
    const PROGRAM_ID: solana_address::Address = EIP191.program_id;
    const SIGNATURE_LEN: usize = 65;
    const IDENTITY_LEN: usize = 20;
    const STORED_IDENTITY_LEN: usize = 20;
}

impl Signer for Eip191 {
    fn identity(&self) -> Vec<u8> {
        let unc = self.key.verifying_key().to_encoded_point(false);
        eth_address_from_pubkey(unc.as_bytes()).to_vec()
    }

    fn sign(&self, digest: &[u8; 32]) -> Vec<u8> {
        let eth = eip191_envelope_hash(digest);
        let (sig, rid): (Signature, RecoveryId) = self.key.sign_prehash(&eth).expect("sign");
        let mut out = vec![0u8; 65];
        out[..64].copy_from_slice(&sig.to_bytes());
        out[64] = rid.to_byte();
        out
    }
}

impl crate::scheme::Registration for Eip191 {}
impl crate::scheme::SingleTxRegister for Eip191 {}

use crate::branching::derive_lane_seed;
impl crate::scheme::Derivable for Eip191 {
    fn derive(&self, index: u32) -> Self {
        let master: [u8; 32] = self.key.to_bytes().into();
        Self::from_seed(&derive_lane_seed(&master, "eip191", index))
    }
}

impl Verifier for Eip191 {
    fn verify(identity: &[u8], _pk: Option<&[u8]>, digest: &[u8; 32], signature: &[u8]) -> bool {
        if signature.len() != 65 {
            return false;
        }
        let eth = eip191_envelope_hash(digest);
        let Ok(sig) = Signature::from_slice(&signature[..64]) else {
            return false;
        };
        let Some(rid) = RecoveryId::from_byte(signature[64]) else {
            return false;
        };
        let Ok(vk) = VerifyingKey::recover_from_prehash(&eth, &sig, rid) else {
            return false;
        };
        let unc = vk.to_encoded_point(false);
        eth_address_from_pubkey(unc.as_bytes()).as_slice() == identity
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::scheme::{Signer, Verifier};

    #[test]
    fn sign_then_verify_roundtrip() {
        let k = Eip191::from_seed(&[8u8; 32]);
        let digest = [4u8; 32];
        let sig = k.sign(&digest);
        assert_eq!(sig.len(), 65);
        assert_eq!(k.identity().len(), 20);
        assert!(Eip191::verify(&k.identity(), None, &digest, &sig));
        let mut bad = sig.clone();
        bad[0] ^= 1;
        assert!(!Eip191::verify(&k.identity(), None, &digest, &bad));
    }
}

File: sdk/ts/src/schemes/eip191.ts
/**
 * secp256k1 + EIP-191 program: identity is the 20-byte Ethereum address;
 * the digest is wrapped in the EIP-191 personal-sign envelope before
 * signing/recovery.
 *
 * Mirrors `crates/core/src/schemes/eip191.rs`. Pulls in only
 * `@noble/curves/secp256k1` + `@noble/hashes/sha3` (keccak).
 */
import { Address, TransactionInstruction } from "@solana/web3.js";
import { secp256k1 } from "@noble/curves/secp256k1";
import { keccak_256 } from "@noble/hashes/sha3";

import { Scheme, findVectorPda } from "../scheme.js";
import {
  createInitializeInstruction,
  createAdvanceInstruction,
} from "../instructions.js";
import { advanceVectorDigest } from "../digest.js";
import { Vector, Artifact } from "../vector.js";
import { ChainSigner, deriveLaneSeed } from "../branching.js";
import { registerArtifactVerifier, artifactParts } from "../inspect.js";

/** secp256k1 ECDSA + EIP-191 envelope — identity is the 20-byte ETH address. */
export const EIP191: Scheme = {
  programId: new Address("G6okL1MvXx7k5eytY7wRXNupXyYG1QVZW37ygAjMiTTu"),
  signatureLen: 65,
  identityLen: 20,
  storedIdentityLen: 20,
};

/** Derive the 20-byte Ethereum address from a 32-byte secp256k1 private key. */
export function ethAddressFromPrivateKey(privateKey: Uint8Array): Uint8Array {
  const pubkey = secp256k1.getPublicKey(privateKey, false); // uncompressed, 65 bytes
  const hash = keccak_256(pubkey.slice(1)); // remove 0x04 prefix
  return hash.slice(12, 32); // last 20 bytes
}

/** EIP-191 identity: the raw 20-byte ETH address. */
export function eip191Identity(privateKey: Uint8Array): Uint8Array {
  return ethAddressFromPrivateKey(privateKey);
}

/** Initialize an EIP-191 vector account. `ethAddress` is the 20-byte address. */
export function createInitializeEip191(
  payer: Address,
  ethAddress: Uint8Array
): TransactionInstruction {
  return createInitializeInstruction(payer, EIP191, ethAddress, ethAddress);
}

/** EIP-191 personal-sign: `keccak256("\x19Ethereum Signed Message:\n32" || digest)` */
const EIP191_PREFIX = new TextEncoder().encode(
  "\x19Ethereum Signed Message:\n32"
);

function eip191Hash(digest: Uint8Array): Uint8Array {
  const buf = new Uint8Array(EIP191_PREFIX.length + digest.length);
  buf.set(EIP191_PREFIX);
  buf.set(digest, EIP191_PREFIX.length);
  return keccak_256(buf);
}

/**
 * Sign the advance digest with an EIP-191 (Ethereum-style) secp256k1 key and
 * return a ready-to-submit advance instruction.
 * @param privateKey 32-byte secp256k1 private key
 */
export function signAdvanceInstructionEip191(
  privateKey: Uint8Array,
  nonce: Uint8Array,
  preInstructions: TransactionInstruction[],
  postInstructions: TransactionInstruction[],
  feePayer?: Address
): TransactionInstruction {
  const identity = eip191Identity(privateKey);
  const digest = advanceVectorDigest(
    EIP191,
    nonce,
    identity,
    preInstructions,
    postInstructions,
    feePayer
  );

  const ethDigest = eip191Hash(digest);
  const sig = secp256k1.sign(ethDigest, privateKey);
  const sigBytes = new Uint8Array(65);
  sigBytes.set(sig.toCompactRawBytes(), 0); // r || s (64 bytes)
  sigBytes[64] = sig.recovery; // v (1 byte)

  return createAdvanceInstruction(EIP191, identity, sigBytes);
}

function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let d = 0;
  for (let i = 0; i < a.length; i++) d |= a[i] ^ b[i];
  return d === 0;
}

/** EIP-191 (Ethereum) chain signer (32-byte secp256k1 private key). */
export function eip191ChainSigner(privateKey: Uint8Array): ChainSigner {
  return {
    scheme: EIP191,
    identity: eip191Identity(privateKey),
    sign: (nonce, pre, post, feePayer) =>
      signAdvanceInstructionEip191(privateKey, nonce, pre, post, feePayer),
  };
}

/** Construct an EIP-191 {@link Vector} from a 32-byte secp256k1 private key. */
export function vectorEip191(
  privateKey: Uint8Array,
  opts?: { feePayer?: Address }
): Vector {
  const signer = eip191ChainSigner(privateKey);
  const [pda] = findVectorPda(EIP191, signer.identity);
  return Vector.fromParts({
    scheme: EIP191,
    signer,
    pda,
    initIx: (payer) => createInitializeEip191(payer, signer.identity),
    feePayer: opts?.feePayer,
    deriveChild: (i) =>
      vectorEip191(deriveLaneSeed(privateKey, "eip191", i), opts),
  });
}

registerArtifactVerifier(EIP191.programId, (a: Artifact) => {
  const parts = artifactParts(a, EIP191);
  if (!parts) return false;
  try {
    const recovered = secp256k1.Signature.fromCompact(parts.signature.slice(0, 64))
      .addRecoveryBit(parts.signature[64])
      .recoverPublicKey(eip191Hash(parts.digest))
      .toRawBytes(false); // 65-byte uncompressed: 0x04 || x || y
    return bytesEqual(keccak_256(recovered.slice(1)).slice(12, 32), a.identity);
  } catch {
    return false;
  }
});
